lib/Backend/Lower.cpp

```diff
@@ -4932,15 +4934,47 @@ Lowerer::LowerNewScObjArray(IR::Instr *newObjInstr)
         Assert(weakFuncRef);
     }
 
-    IR::Opnd *opndSrc1 = newObjInstr->UnlinkSrc1();
-    if (opndSrc1->IsImmediateOpnd())
+    IR::LabelInstr *labelDone = IR::LabelInstr::New(Js::OpCode::Label, func);
+    IR::Opnd *opndSrc = newObjInstr->GetSrc2();
+
+    Assert(opndSrc->IsSymOpnd());
+    StackSym *linkSym = opndSrc->AsSymOpnd()->m_sym->AsStackSym();
+    Assert(linkSym->IsSingleDef());
+    IR::Instr* argInstr = linkSym->GetInstrDef();
+    IR::Opnd *opndOfArrayCtor = argInstr->GetSrc1();
+    // Generate fast path only if it meets all the conditions:
+    // 1. 1st paramter is a constant
+    // 2. It is in range 0 <= parameter <= 8
+    // 3. It is the only parameter.
+    if (opndOfArrayCtor->IsIntConstOpnd() || opndOfArrayCtor->IsAddrOpnd()) // #1
     {
-        intptr_t length = opndSrc1->GetImmediateValue();
-        if (length >= 0 && length <= 8)
+        int32 length = linkSym->GetIntConstValue();
+
+        if (length >= 0 && length <= 8) // #2
         {
-            GenerateProfiledNewScObjArrayFastPath(newObjInstr, arrayInfo, weakFuncRef, (uint32)length);
+            AssertMsg(linkSym->IsArgSlotSym(), "Not an argSlot symbol...");
+            opndSrc = argInstr->GetSrc2();
+
+            // If more than 1 args, then opndSrc would be symOpnd
+            bool emitFastPath = !opndSrc->IsSymOpnd(); // #3
```

Comment:
Still this is kinda multiple implicit assumptions about the operands - instead if you rely on the slotNum, it will be more explicit and more straightforward about whether you need a fast path or not.

---
In reply to: [71741291](https://github.com/Microsoft/ChakraCore/pull/1304#discussion_r71741291) [](ancestors = 71741291,71649040)